### PR TITLE
MGMT-9357 split api dir into a sub module

### PR DESCRIPTION
**Pull request by filanov**
It will enable to import api without involving in the main module replace
that we have in go.mod files.

Update golang version to 1.17 and update go.mod to golang 1.17
It will prevent some of the replaces that we need in go.mod file
Specifically resolving a conflict in sigs.k8s.io/json that uses
reflect package features from golang 1.17

Made go.mod to be compatible for both golang 1.16 and 1.17 it increase
the size of go.mod, it i selected to support only golang 1.17 go.mod
will size still increased so didn't found any issue supporting both

Fix code issues related to new k8s packages versions

Update hive and metal3 crds after version update

When generating manifests the command need to be executed from the api
directory, otherwise controller-gen will not set the sub modules and
will not generate anything even if directly passing ./api/... in the
paths parameter

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @flaper87 
/cc @osherdp 

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by filanov** [comment]
/hold 
@osherdp is there a reason we didn't moved to golang 1.17?
If not then is there anything else that need to change in order to move everything to 1.17? downstream builds or any other place?

**Comment by omertuc** [comment]
Here is all the things I had to do for 14/15 -> 16, if you want to do the same for 1.17... 

https://issues.redhat.com/browse/MGMT-7210
https://code.engineering.redhat.com/gerrit/c/assisted-installer-projects/+/253860
https://github.com/openshift/release/pull/20147
https://github.com/openshift/release/pull/20146
https://github.com/openshift/release/pull/20145
https://github.com/openshift/assisted-service/pull/2203
https://github.com/openshift/assisted-installer-agent/pull/215
https://github.com/openshift/assisted-installer/pull/325

**Comment by filanov** [comment]
release pr https://github.com/openshift/release/pull/26599

**Comment by filanov** [comment]
there is an infra issue in test-infra. waiting for a fix.

**Comment by osherdp** [comment]
/retest
(assisted-test-infra image promotion took a bit more time to complete)

**Comment by osherdp** [comment]
```
/go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:2760
Test Panicked
/usr/local/go/src/runtime/panic.go:221

Panic: runtime error: invalid memory address or nil pointer dereference

Full stack:
github.com/openshift/assisted-service/subsystem.glob..func33.38()
	/go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:2895 +0x1994
github.com/openshift/assisted-service/subsystem.TestSubsystem(0xc000ad5301)
	/go/src/github.com/openshift/assisted-service/subsystem/subsystem_suite_test.go:161 +0xb7
testing.tRunner(0xc000664820, 0x1f7b170)
	/usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1306 +0x35a
```

Seems specific for this PR
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/3406/pull-ci-openshift-assisted-service-master-subsystem-kubeapi-aws/1497891955059724288

File: api/go.mod [hunk 1]
```diff
@@ -0,0 +1,50 @@
+module github.com/openshift/assisted-service/api
+
+go 1.17
+
+require (
+	github.com/openshift/assisted-service v1.0.9
```
Comment (osherdp) [inline api/go.mod:6]: api module depends on the "root" module?

**Review by osherdp — CHANGES_REQUESTED** [review]
One thing we should make sure here, is to locally verify the Dockerfile changes (or be prepared to handle it when it merges)
That's because some of those Dockerfile changes are not reflected in the jobs in prow, and it rather takes those files from master branch

There's a way to change this behavior if we want to, at least in test-infra it worked pretty good (this repo is more complex though, so no idea how it will behave here)

File: api/go.mod (hunk 1 above)
Comment (eranco74) [reply api/go.mod:6]: This seems wrong

Comment (filanov) [reply api/go.mod:6]: Sorry you are right. strange how it worked. thanks will fix

**Review by osherdp — APPROVED** [review]

**Comment by filanov** [comment]
> One thing we should make sure here, is to locally verify the Dockerfile changes (or be prepared to handle it when it merges) That's because some of those Dockerfile changes are not reflected in the jobs in prow, and it rather takes those files from master branch
> 
> There's a way to change this behavior if we want to, at least in test-infra it worked pretty good (this repo is more complex though, so no idea how it will behave here)

I tried update-minimal it checks the the build. anything else that i should check?

**Comment by filanov** [comment]
> ```
> /go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:2760
> Test Panicked
> /usr/local/go/src/runtime/panic.go:221
> 
> Panic: runtime error: invalid memory address or nil pointer dereference
> 
> Full stack:
> github.com/openshift/assisted-service/subsystem.glob..func33.38()
> 	/go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:2895 +0x1994
> github.com/openshift/assisted-service/subsystem.TestSubsystem(0xc000ad5301)
> 	/go/src/github.com/openshift/assisted-service/subsystem/subsystem_suite_test.go:161 +0xb7
> testing.tRunner(0xc000664820, 0x1f7b170)
> 	/usr/local/go/src/testing/testing.go:1259 +0x102
> created by testing.(*T).Run
> 	/usr/local/go/src/testing/testing.go:1306 +0x35a
> ```
> 
> Seems specific for this PR https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/3406/pull-ci-openshift-assisted-service-master-subsystem-kubeapi-aws/1497891955059724288

a very strange error, it didn't manage to get the secret but when trying to create it's getting an error
```
time="2022-02-27T15:33:17Z" level=error msg="failed to create kubeconfig no-ingress secret" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:226" agent_cluster_install=test-agent-cluster-install-7f0af670 agent_cluster_install_namespace=ci-op-b7ysj89g cluster_deployment=test-cluster-128a43b3 cluster_deployment_namespace=ci-op-b7ysj89g error="secrets \"test-cluster-128a43b3-admin-kubeconfig\" already exists" go-id=675 request_id=6e2a3fab-e29a-4a9d-99e4-f4ed64bb5ce7
```

**Comment by osherdp** [comment]
> > One thing we should make sure here, is to locally verify the Dockerfile changes (or be prepared to handle it when it merges) That's because some of those Dockerfile changes are not reflected in the jobs in prow, and it rather takes those files from master branch
> > There's a way to change this behavior if we want to, at least in test-infra it worked pretty good (this repo is more complex though, so no idea how it will behave here)
> 
> I tried update-minimal it checks the the build. anything else that i should check?

I want to say ``Dockerfile.bundle`` but I think it's actually unused

File: go.mod [hunk 2]
```diff
@@ -69,35 +68,183 @@ require (
...
+	sigs.k8s.io/cluster-api-provider-aws v0.0.0-00010101000000-000000000000 // indirect
+	sigs.k8s.io/cluster-api-provider-azure v0.0.0-00010101000000-000000000000 // indirect
+	sigs.k8s.io/json v0.0.0-20211020170558-c049b76a60c6 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
+)
+
 replace (
-	github.com/googleapis/gnostic => github.com/googleapis/gnostic v0.4.1
-	github.com/metal3-io/baremetal-operator => github.com/openshift/baremetal-operator v0.0.0-20210409032903-31b989a197eb // Use OpenShift fork
+	github.com/metal3-io/baremetal-operator/apis => github.com/openshift/baremetal-operator/apis v0.0.0-20220217140404-6b1ecb71984f
```
Comment (flaper87) [inline go.mod:243]: Nice! I had to do this for my convergence PoC, good to get this in asap.

**Review by flaper87 — COMMENTED** [review]
looks good to me, just need to figure out the failures (stating the obvious, I know :smile:  )

**Comment by filanov** [comment]
I think that i found the issue with the tests. looks like hive and metal3 did some changes in the crds. 
We keep the crds locally and deploy then for the tests. the code had a new version but the definition was old. 
Had to update the crds as well.

**Review by osherdp — APPROVED** [review]

**Comment by filanov** [comment]
same error as before
```
time="2022-02-28T08:00:54Z" level=error msg="failed to create kubeconfig no-ingress secret" func="github.com/openshift/assisted-service/internal/controller/controllers.(*ClusterDeploymentsReconciler).Reconcile" file="/go/src/github.com/openshift/origin/internal/controller/controllers/clusterdeployments_controller.go:226" agent_cluster_install=test-agent-cluster-install-8831f6bd agent_cluster_install_namespace=ci-op-gl985fvj cluster_deployment=test-cluster-f28d60ed cluster_deployment_namespace=ci-op-gl985fvj error="AgentClusterInstall.extensions.hive.openshift.io \"test-agent-cluster-install-8831f6bd\" is invalid: spec.clusterMetadata.adminPasswordSecretRef: Required value" go-id=560 request_id=37d2cbc3-65d4-426e-b11d-c68ead918e80
```
It indicate that the crd is not updated, does it take it from master or something like that?

**Comment by filanov** [comment]
interesting problems. splitting the sub modules broke the manifests generation flow. because it's a sub module `controller-gen` ignored it in the path, even after i explicitly added it. it's similar to `go list` for example
```
go list ./... ./api...
go: warning: "./api..." matched no packages
...
```
vs
```
cd api/ ; go list ./...
github.com/openshift/assisted-service/api/common
github.com/openshift/assisted-service/api/hiveextension/v1beta1
github.com/openshift/assisted-service/api/v1beta1
```
So i got to cd into `api` and then run the generation

**Comment by dtantsur** [comment]
I could successfully resolve the dependencies now, great: https://github.com/dtantsur/image-customization-controller/commit/99e8065836e27865594b3f4da3efe2a329b131d7

**Comment by filanov** [comment]
Another interesting issue (thanks @carbonin for helping)
k8s packages changed some of the behavior between the versions, looks
like right now update will return a new copy of the object
In our case Agent reconcile make Annotations and Labels maps if nil,
when adding a finalizer the code need to stop the flow and reconcile
again and not try to use the new object when the maps are nil again. It
caused panic in the service

File: go.mod [hunk 3]
```diff
@@ -69,35 +68,183 @@ require (
...
+	k8s.io/apiserver v0.23.0 // indirect
+	k8s.io/component-base v0.23.0 // indirect
+	k8s.io/klog/v2 v2.40.1 // indirect
+	k8s.io/kube-openapi v0.0.0-20220124234850-424119656bbf // indirect
+	sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.25 // indirect
+	sigs.k8s.io/cluster-api-provider-aws v0.0.0-00010101000000-000000000000 // indirect
+	sigs.k8s.io/cluster-api-provider-azure v0.0.0-00010101000000-000000000000 // indirect
+	sigs.k8s.io/json v0.0.0-20211020170558-c049b76a60c6 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
+)
```
Comment (carbonin) [inline go.mod:240]: What is this?

**Comment by carbonin** [comment]
> k8s packages changed some of the behavior between the versions

Were you able to figure out what change this was?

File: go.mod (hunk 3 above)
Comment (filanov) [reply go.mod:240]: Dependencies that 1.17 added. regardless of the support for 1.16. 
i didn't manage to find how to remove it. 
go mod tidy (form inside the build container) added it.

**Comment by filanov** [comment]
> > k8s packages changed some of the behavior between the versions
> 
> Were you able to figure out what change this was?

I didn't find any documentation about it.
And the code looks complicated i didn't see any changes in the `zz_*` files 
So i'm not really sure

**Comment by filanov** [comment]
Looks like infra error
/test subsystem-kubeapi-aws

**Comment by filanov** [comment]
ztp is flaky

**Comment by filanov** [comment]
A side effect of the packages update it looks like when running the webhooks sever they removed the log level flag, i didn't managed to find the commit or the reason but i complied the webhooks locally and it was removed
```
./build/from-master -h | grep "\--v"
  -v, --v Level                                                 number for the log level verbosity
      --vmodule moduleSpec                                      comma-separated list of pattern=N settings for file-filtered logging
```
vs
```
./build/sub-modules -h | grep "\--v"
```
it crashed the webhooks pod in the ztp test

File: apiserver.local.config/certificates/apiserver.key [hunk 4]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (osherdp) [inline apiserver.local.config/certificates/apiserver.key:1]: where do you use those certificates?

**Review by osherdp — CHANGES_REQUESTED** [review]

File: apiserver.local.config/certificates/apiserver.key (hunk 4 above)
Comment (osherdp) [reply apiserver.local.config/certificates/apiserver.key:1]: also why not generating them on the fly if they're used for tests

Comment (filanov) [reply apiserver.local.config/certificates/apiserver.key]: It was generated when i tried to run locally the admission hooks binary i just didn't see that it was added, related to https://github.com/openshift/assisted-service/pull/3406#issuecomment-1055431677 . 
good catch, removed.

**Comment by filanov** [comment]
/retest 
Looks like infra issues

**Review by osherdp — APPROVED** [review]

**Comment by osherdp** [comment]
/hold
let's wait for ztp to get fixed: https://github.com/openshift-metal3/dev-scripts/pull/1356

**Comment by osherdp** [comment]
> /hold let's wait for ztp to get fixed: [openshift-metal3/dev-scripts#1356](https://github.com/openshift-metal3/dev-scripts/pull/1356)

/unhold
fixed

**Comment by filanov** [comment]
/test e2e-metal-assisted-operator-ztp
```
2022-03-02 18:06:28 fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: list object has no element 0\n\nThe error appears to be in '/output/packet-setup.yaml': line 34, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n      no_log: true\n    - name: wait for ssh\n      ^ here\n"}
2022-03-02 18:06:28 
2022-03-02 18:06:28 TASK [Send notification message via Slack in case of failure] ******************
2022-03-02 18:06:29 ok: [localhost]
2022-03-02 18:06:29 
2022-03-02 18:06:29 TASK [fail the play] ***********************************************************
2022-03-02 18:06:29 fatal: [localhost]: FAILED! => {"changed": false, "msg": "ERROR: Packet setup failed."}
```

**Pull request merged**